### PR TITLE
assistant deploy

### PR DESCRIPTION
**Pull request by ThWoywod**
Adds the CLIDeploymentExtension implementation for assistant-alexa.
If the cli command `assistant deploy` will be called, the currently generated configuration will be uploaded to the *Alexa Developer Console*.

### CODE REVIEW

File: src/components/alexa/deployment.ts [hunk 1]
```diff
@@ -0,0 +1,310 @@
+import { CLIDeploymentExtension, injectionNames, Logger } from "assistant-source";
+import { execSync } from "child_process";
+import * as fs from "fs";
+import { inject, injectable } from "inversify";
+import { Component, getMetaInjectionName } from "inversify-components";
+import { AlexaSkillSchema, COMPONENT_NAME, Configuration } from "./private-interfaces";
+
+@injectable()
+export class AlexaDeployment implements CLIDeploymentExtension {
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: JSDoc

Comment (baflo) [inline src/components/alexa/deployment.ts]: JSDoc

File: src/components/alexa/deployment.ts [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+  constructor(
+    @inject(getMetaInjectionName(COMPONENT_NAME)) private componentMeta: Component<Configuration.Runtime>,
+    @inject(injectionNames.logger) private logger: Logger
+  ) {}
+
+  public async execute(buildPath: string) {
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: JSDoc

File: src/components/alexa/deployment.ts [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+      this.deploySkillSchema(buildPath, countryCodes);
+
+      await Promise.all(
+        countryCodes.map(async countryCode => {
+          const currentLocale = languageMapping(countryCode);
+          await this.exportModel(buildPath, currentLocale);
+          await this.updateModel(buildPath, countryCode, currentLocale);
+
+          // Wait until the model upload is out of state in progress.
+          await this.whileModelIsInProgress(countryCode);
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: This is not exceptionally clear. I suppose it checks whether agent has finished training the updated skill. Comments? Links to documentation?

Maybe rename the function to sth like `whileModelTrainingIsInProgress`

File: src/components/alexa/deployment.ts [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+  private async getModel(locale: string) {
+    const modelSchema = execSync(`ask api get-model -s ${this.componentMeta.configuration.applicationID} -l ${locale}`);
+    return JSON.parse(modelSchema.toString());
+  }
+
+  private async updateModel(buildPath: string, countryCode: string, locale: string) {
+    try {
+      // Execute the ask update model command. e.g. 'ask api update-model -s schema.json -l de-DE'
+      const updateModelExecution = execSync(
+        `ask api update-model -s ${this.componentMeta.configuration.applicationID} -f ${buildPath}/${countryCode}/alexa/schema.json -l ${locale}`
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: `path.join`

Comment (baflo) [inline src/components/alexa/deployment.ts]: `path.join`

File: src/components/alexa/deployment.ts [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+  /**
+   * Export the currently deployed alexa model and store it to the given buildPath
+   * @param buildPath Path to the current build folder. like {root}/builds/123456789
+   * @param locale LCID Code
+   */
+  private async exportModel(buildPath: string, locale: string) {
+    const model = await this.getModel(locale);
+    try {
+      fs.writeFileSync(`${buildPath}/deployments/alexa/schema_${locale}.json`, JSON.stringify(model, null, 2));
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: `path.join`

File: src/components/alexa/deployment.ts [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+  public async execute(buildPath: string) {
+    // tslint:disable-next-line:no-console
+    console.log("===============     APIAI DEPLOYMENT     ===============");
+    // Reads all given languages from folder structure.
+    const countryCodes = fs.readdirSync(`${buildPath}`);
+    if (!countryCodes || countryCodes.length === 0) {
+      throw new Error("There is no configuration given: Please execute the 'assistant generator' before uploading the current configuration.");
+    }
+    if (this.isAskInstalled()) {
+      this.deploySkillSchema(buildPath, countryCodes);
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: Below you divide export and deploy (which's named "update" there), but here export is a side effect of deploy. Stick to a common convention if possible

File: src/components/alexa/deployment.ts [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+  private logModelBuildStatus(countryCode: string) {
+    // tslint:disable-next-line:no-console
+    console.log(`Amazon model building for ${languageMapping(countryCode)}: ${this.status(countryCode)}`);
+  }
+
+  /**
+   *  Wait until the status of the model upload is in state IN_PROGRESS
+   * @param countryCode Country code like de or en
+   */
+  private async whileModelIsInProgress(countryCode: string) {
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: what's the model progressing?

File: src/components/alexa/deployment.ts [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+  /**
+   * Get the current alexa skill configuration from disk.
+   * @param buildPath Path to the current build folder like 'builds/1550067893551'
+   * @returns skill schema as @type {AlexaSkillSchema}
+   */
+  private getCurrentSkillSchema(buildPath: string) {
+    if (!fs.existsSync(`${buildPath}/deployments/alexa/skill-backup.json`)) {
+      this.exportCurrentSkillSchema(buildPath);
+    }
+    const alexaSkillSchema: AlexaSkillSchema = JSON.parse(fs.readFileSync(`${buildPath}/deployments/alexa/skill-backup.json`).toString());
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: `path.join`

File: src/components/alexa/deployment.ts [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+    return alexaSkillSchema;
+  }
+
+  /**
+   * Generates the locales schema form the given country codes
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: typo
> form

File: src/components/alexa/deployment.ts [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+  }
+
+  /**
+   * Update the alexa skill configuration.
+   * @param buildPath Path to the current build folder like 'builds/1550067893551'
+   * @param countryCodes Array of country codes like ["de", "en"]
+   * @param skillSchema @type {AlexaSkillSchema} witch should be updated
+   */
+  private async updateSkillSchema(buildPath: string, countryCodes: string[], skillSchema: AlexaSkillSchema) {
+    fs.writeFileSync(`${buildPath}/deployments/alexa/skill.json`, JSON.stringify(skillSchema, null, 2));
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: `path.join`

File: src/components/alexa/deployment.ts [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+/**
+ * Alexa needs the locales in a LCID Format. Currently we use the country code for indicating the language.
+ * This mapping function allows you to get the LCID Code for the country codes: de or en.
+ * If an unknown country code will be given, these will not be mapped and the given country code will be returned.
+ * @param countryCode country code as a string like de or en
+ * @returns LCID Code or country code as a string
+ */
+const languageMapping = countryCode => {
```
Comment (baflo) [inline src/components/alexa/deployment.ts]: add as private member to class

File: spec/src/components/alexa/deployment.spec.ts [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+    /** Replace console log and error with an spy instance */
+    (console as any).error = jasmine.createSpy("console.error");
+    (console as any).log = jasmine.createSpy("console.log");
+
+    /**
+     * Mock the setInterval method used by the AlexaDeployment
+     */
+    (setInterval as any) = jasmine.createSpy("setInterval").and.callFake((callback: () => {}, intervalTime: number) => {
+      // setInterval is an asynchronous function we first have to return a value before the callback method will be invoked
+      setTimeout(callback.bind(this.alexaDeployment), 0);
```
[the remark this answers is not in the document]
Comment (ThWoywod) [reply spec/src/components/alexa/deployment.spec.ts]: Yes thats right, but if i use `process.nextTick` some exception will be thrown. I try to fix it but currently i do not have any idea why this error happen.

```
  this.execSyncSpy = (execSync as any) = jasmine.createSpy("execSync").and.callFake((command: string) => { ...})
  ...
  (setInterval as any) = jasmine.createSpy("setInterval").and.callFake((callback: () => {}, intervalTime: number) => {
      process.nextTick(callback.bind(this.alexaDeployment));
      return 1;
  });
```

It looks like, that the `callback`-Function do not know the mocked instances of `execSync` Function.

File: spec/src/components/alexa/deployment.spec.ts [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+                `${this.buildPath}/deployments/alexa/skill-backup.json`,
+                this.spyReturns.execSync.getSkill.toString()
+              );
+            });
+          });
+
+          describe("regarding skill schema generation ", function() {
+            describe("with new locales", function() {
+              beforeEach(async function(this: CurrentThisContext) {
+                // Mock exported Skill schema
```
[the remark this answers is not in the document]
Comment (ThWoywod) [reply spec/src/components/alexa/deployment.spec.ts]: In `assistant-apai` it's not needed to update the skill schema. During the restore process Dialogflow overwrites the agent configuration by default. So you Agent always use the AssistantJS generated configuration.

Alexa do not have this feature. You have to configure the agent by your own with its own api. So for example if your agent should handle multiple languages you have first update the agent configuration and then update the skill schema.

[the remark this answers is not in the document]
Comment (baflo) [reply spec/src/components/alexa/deployment.spec.ts]: Ah, I see... I didnt't recognize the difference between schema and model. Any chance you can make that more clear?

[the remark this answers is not in the document]
Comment (baflo) [reply spec/src/components/alexa/deployment.spec.ts]: I mean, can you make that more clear in `deployment.ts`?

**Pull request merged**